### PR TITLE
Add DragonFly bootstrap support (ltsmaster) (/src/core/sys/posix)

### PR DESCRIPTION
**Pull request by dkgroot**
All druntime:/src/core/sys/posix/ files

Related PR's:
- ldc
  - [ldc:ltsmaster](https://github.com/ldc-developers/ldc/pull/2461)
  - druntime:ltsmaster
    - [base PR](https://github.com/ldc-developers/druntime/pull/109)
    - [/src/core/sys/posix](https://github.com/ldc-developers/druntime/pull/111)
    - [/src/core/sys/dragonflybsd](https://github.com/ldc-developers/druntime/pull/110)
  - [phobos:ltsmaster](https://github.com/ldc-developers/phobos/pull/45)
  - [dmd-testsuite:ltsmaster](https://github.com/ldc-developers/dmd-testsuite/pull/28)

- dlang:
  - [dlang/dmd:master](https://github.com/dlang/dmd/pull/7463)
  - dlang/druntime:master
    - [dlang/druntime:master](https://github.com/dlang/druntime/pull/1999)
    - [dlang/druntime:master/src/core/sys/dragonflybsd](https://github.com/dlang/druntime/pull/2002)
    - [dlang/druntime:master/src/core/sys/posix](https://github.com/dlang/druntime/pull/2001)
  - [dlang/phobos:master](https://github.com/dlang/phobos/pull/5950) (merged already)

### CODE REVIEW

**Comment by joakim-noah** [comment]
Go ahead and update this to the latest version that got merged upstream and I'll pull it.

**Comment by dkgroot** [comment]
Rebase and synced with dlang/PR/2001

**Comment by dkgroot** [comment]
@joakim-noah : Rebase and synced with dlang/PR/2001
Ready to Merge
[DFly SemaphoreCI](https://semaphoreci.com/dkgroot/ldc_dragonfly_ci)

**Comment by dkgroot** [comment]
@joakim-noah 
[math.d] I need to either use all the platform/os specific version of the math functions, or use the llvm builtin ones. Mixing them leads to problems with rounding during asserts.
[Trivial] The pthread_np.d issues (see CI tests) are caused by the filename and module name not matching up.

File: src/core/sys/posix/pthread.d [hunk 1]
```diff
@@ -209,6 +209,46 @@ else version( FreeBSD )
...
+    }
+
+    enum PTHREAD_CANCELED = cast(void*) -1;
+
+    enum PTHREAD_NEEDS_INIT = 0;
+    enum PTHREAD_DONE_INIT  = 1;
+
+    enum PTHREAD_MUTEX_INITIALIZER              = null;
+    //enum PTHREAD_ONCE_INIT                      = { PTHREAD_NEEDS_INIT, NULL };
+    enum PTHREAD_ONCE_INIT                      = pthread_once_t.init;;
```
Comment (joakim-noah) [inline src/core/sys/posix/pthread.d]: Delete one semi-colon, missed this upstream too.

File: src/core/sys/posix/pthread.d [hunk 2]
```diff
@@ -648,6 +738,25 @@ else version( FreeBSD )
...
     int pthread_setconcurrency(int);
 }
+else version( DragonFlyBSD )
+{
+    enum
+    {
+        PTHREAD_MUTEX_ERRORCHECK    = 1,
+        PTHREAD_MUTEX_RECURSIVE     = 2,
+        PTHREAD_MUTEX_NORMAL        = 3,
+        PTHREAD_MUTEX_ADAPTIVE_NP   = 4,
```
Comment (joakim-noah) [inline src/core/sys/posix/pthread.d]: This isn't there upstream.

File: src/core/sys/posix/sys/shm.d [hunk 3]
```diff
@@ -121,6 +121,31 @@ else version( FreeBSD )
     int   shmdt(in void*);
     int   shmget(key_t, size_t, int);
 }
+else version( DragonFlyBSD )
```
Comment (joakim-noah) [inline src/core/sys/posix/sys/shm.d:124]: Neither is this.

**Review by joakim-noah — CHANGES_REQUESTED** [review]
Some questions.

File: src/core/sys/posix/pthread.d (hunk 1 above)
Comment (dkgroot) [reply src/core/sys/posix/pthread.d]: Done

File: src/core/sys/posix/pthread.d (hunk 2 above)
Comment (dkgroot) [reply src/core/sys/posix/pthread.d]: Removed ```PTHREAD_MUTEX_ADAPTIVE_NP   = 4,```
Thanks !

File: src/core/sys/posix/sys/shm.d (hunk 3 above)
Comment (dkgroot) [reply src/core/sys/posix/sys/shm.d:124]: Ooh... but it should be (will amend dlang/druntime). Thanks for noticing this one !

Comment (dkgroot) [reply src/core/sys/posix/sys/shm.d]: Link: https://github.com/dlang/druntime/pull/2106

Comment (dkgroot) [reply src/core/sys/posix/sys/shm.d:124]: dlang#2106 has been approved => waiting to be merged

**Comment by dkgroot** [comment]
@joakim-noah Thanks for review and merging !

**Pull request merged**